fem/gridfunc.hpp

```diff
@@ -854,6 +855,60 @@ class ExtrudeCoefficient : public Coefficient
 GridFunction *Extrude1DGridFunction(Mesh *mesh, Mesh *mesh2d,
                                     GridFunction *sol, const int ny);
 
+/// The class provides methods for projecting function values evaluated on a
+/// set of points to a grid function. The values are directly copied to the
+/// nodal values of the target grid function if any of the points is matching
+/// a node of the grid function.
+class KDTreeNodalProjection
+{
+private:
+
+   /// Pointer to the KDTree for the 3D case
+   KDTree3D* kdt3D;
+
+   /// Pointer to the KDTree for the 2D case
+   KDTree2D* kdt2D;
```

Comment:
Since the `KDTree` class is already templated, why not just make this class a template class on the dimension and eliminate the need to have both `kdt3D` and `kdt2D` members when only one is ever non-`NULL`?

Reply:
What about using `::std::unique_ptr` instead of a raw pointer?

Reply:
I think doubling the pointer is a low price to pay for the generality. If we template the projection class - it will be inconvenient for the users. 

To me, smart pointers make the code unreadable and since we have only two raw pointers I would prefer to keep them raw.

Reply:
Why not use typedefs in the same way as for `KDTree3D`?

This would be much clearer to me as:

```
template <size_t ndim=3>
class KDTreeNodalProjection
{
private:
   KDTree<int,double,ndim> kdt;
   Vector minbb, maxbb;

public:
    KDTreeNodalProjection(FiniteElementSpace &fes);

   void Project(const Vector &coords,
               const Vector &src,
               int ordering,
               GridFunction &dest,
               double lerr = 1e-8);

   void Project(const Vector &coords,
               const GridFunction &src,
               GridFunction &dest,
               double lerr = 1e-8);
}

typedef KDTreeNodalProjection<3> KDTreeNodalProjection3D;
typedef KDTreeNodalProjection<2> KDTreeNodalProjection2D;
```

The target `GridFunction` `dest` makes more sense as an argument to `Project` than a member of the class.

Reply:
> To me, smart pointers make the code unreadable and since we have only two raw pointers I would prefer to keep them raw.

It's only the declaration that changes, and ::std:;unique_ptr is so much safer than raw pointers.

Reply:
The KDTree is built around the dest grid function. We do not want to build the tree every time for a new projection - therefore dest should be member of the class.

The suggested templating will work fine, however, MFEM examples/miniapps are often written in dimension independent fashion. So if we define:
typedef KDTreeNodalProjection<3> KDTreeNodalProjection3D;
typedef KDTreeNodalProjection<2> KDTreeNodalProjection2D;

We will need additional if statement is examples/miniapps to check for the dimensionality of the space. The current implementation will work for 2D and 3D spaces without any changes.

Reply:
changed to unique_ptr :-)